src/widgets/Console.ts

```diff
@@ -0,0 +1,91 @@
+import { v, w } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+import { WidgetProperties } from '@dojo/widget-core/interfaces';
+import { ThemedMixin, ThemedProperties, theme } from '@dojo/widget-core/mixins/Themed';
+import { stringify } from '../util';
+import ActionBar, { ActionBarButton } from './ActionBar';
+import * as css from '../styles/console.m.css';
+
+export enum ConsoleMessageType {
+	Log = 'log',
+	Error = 'error',
+	Info = 'info',
+	Warning = 'warn'
+}
+
+export interface ConsoleMessage {
+	type: ConsoleMessageType;
+	message: any | any[];
+	filename?: string;
+	lineNumber?: number;
+}
+
+export interface ConsoleProperties extends WidgetProperties, ThemedProperties {
+	messages?: ConsoleMessage[];
+	onClear?(): void;
+}
+
+function getLogType(type: ConsoleMessageType): string {
+	return type === ConsoleMessageType.Log
+		? css.log : type === ConsoleMessageType.Error
+		? css.error : type === ConsoleMessageType.Info
+		? css.info : css.warn;
+}
+
+const ThemedBase = ThemedMixin(WidgetBase);
+
+export interface ConsoleRowProperties extends WidgetProperties, ThemedProperties {
+	key: string;
+	message: ConsoleMessage;
+}
+
+@theme(css)
+export class ConsoleRow extends ThemedBase<ConsoleRowProperties> {
+	render() {
+		const { key, message: { message, type } } = this.properties;
+		const args = Array.isArray(message) ? message : [ message ];
+		return v('div', {
+			key,
+			classes: this.theme([ getLogType(type), css.consoleRow ])
+		}, [
+			v('span', {
+				classes: this.theme(css.meta)
+			}, [ v('span', { classes: this.theme(css.typeLabel) }, [ type]) ]),
+			v('span', {
+				classes: this.theme(css.logBody)
+			}, [ ...args.map((arg) => v('span', { classes: this.theme(css.logArg) }, [ stringify(arg) ])) ])
```

Comment:
You _always_ map `args` here when rendering it as well as then you are spreading a single array into an array literal... all which seems unnecessary and redundant.  Couldn't you just at 46 create a `const children` in a done and one and not spread the array here?